src/celeritas/io/EventWriter.cc

```diff
@@ -0,0 +1,205 @@
+#include <HepMC3/WriterHEPEVT.h>
+
+#include "corecel/io/EnumStringMapper.hh"
+#include "corecel/io/Join.hh"
+#include "corecel/io/Logger.hh"
+#include "corecel/io/ScopedTimeAndRedirect.hh"
+#include "corecel/io/StringEnumMapper.hh"
+#include "celeritas/phys/ParticleParams.hh"
+#include "celeritas/phys/Primary.hh"
+
+namespace celeritas
+{
+namespace
+{
+//---------------------------------------------------------------------------//
+EventWriter::Format filename_to_format(std::string const& filename)
+{
+    std::string ext;
+    auto pos = filename.rfind('.');
+    if (pos != std::string::npos)
+    {
+        ext = filename.substr(pos + 1);
+    }
+
+    static auto const from_string
+        = StringEnumMapper<EventWriter::Format>::from_cstring_func(
+            to_cstring, "event file format");
+    return from_string(ext);
+}
+
+//---------------------------------------------------------------------------//
+}  // namespace
+
+//---------------------------------------------------------------------------//
+/*!
+ * Construct by parsing the extension.
+ */
+EventWriter::EventWriter(std::string const& filename, SPConstParticles params)
+    : EventWriter{filename, std::move(params), filename_to_format(filename)}
+{
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Construct with a filename, particle data, and output format.
+ */
+EventWriter::EventWriter(std::string const& filename,
+                         SPConstParticles particles,
+                         Format fmt)
+    : particles_{particles}, fmt_{fmt}
+{
+    CELER_EXPECT(!filename.empty());
+    CELER_EXPECT(particles_);
+    CELER_EXPECT(fmt_ != Format::size_);
+
+    CELER_LOG(info) << "Creating " << to_cstring(fmt) << " event file at "
+                    << filename;
+    ScopedTimeAndRedirect temp_{"HepMC3"};
+
+    writer_ = [&]() -> std::shared_ptr<HepMC3::Writer> {
+        switch (fmt)
+        {
+            case Format::hepevt:
+                return std::make_shared<HepMC3::WriterHEPEVT>(filename);
+            case Format::hepmc2:
+                return std::make_shared<HepMC3::WriterAsciiHepMC2>(filename);
+            case Format::hepmc3:
+                return std::make_shared<HepMC3::WriterAscii>(filename);
+            default:
+                CELER_ASSERT_UNREACHABLE();
+        }
+    }();
+
+    CELER_ENSURE(writer_);
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Write all primaries from a single event.
+ */
+void EventWriter::operator()(argument_type primaries)
+{
+    CELER_EXPECT(!primaries.empty());
+
+    std::set<EventId::size_type> mismatched_events;
+
+    HepMC3::GenEvent evt(HepMC3::Units::MEV, HepMC3::Units::CM);
+
+    // Vertex and corresponding celeritas position
+    HepMC3::GenVertexPtr vtx;
+    Primary const* vtx_primary{nullptr};
+
+    // See HepMC2 manual
+    enum HepStatus
+    {
+        test_status = 0,
```

Comment:
It doesn't look like this is used anywhere.

Reply:
I've added all the status codes so I don't have to go looking through the reference manual every time 😉